go/services/connector/main.go

```diff
@@ -0,0 +1,620 @@
+		m.status = ConnectionStatusDisconnected
+		m.statusChange.Broadcast()
+	}
+}
+
+// WaitForConnected waits until the connection is connected.
+func (m *ConnectionMonitor) WaitForConnected(ctx context.Context) error {
+	done := make(chan struct{})
+
+	go func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		for m.status != ConnectionStatusConnected {
+			m.statusChange.Wait()
+		}
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// WaitForDisconnected waits until the connection is disconnected.
+func (m *ConnectionMonitor) WaitForDisconnected(ctx context.Context) error {
+	done := make(chan struct{})
+
+	go func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		for m.status != ConnectionStatusDisconnected {
+			m.statusChange.Wait()
+		}
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// DeviceEndpointCreateObservation represents an observation for device endpoint creation events.
+type DeviceEndpointCreateObservation struct {
+	mountPath    string
+	deviceChan   chan DeviceNotification
+	fileMountMap *FileMountMap
+	ctx          context.Context
+	cancel       context.CancelFunc
+	pollInterval time.Duration // retained for API compatibility; not used
+	connMonitor  *ConnectionMonitor
+	logger       *slog.Logger
+
+	watcher     *fsnotify.Watcher
+	lastDevices map[string]DeviceEndpointRef
+}
+
+// DeviceNotification contains information about a device endpoint notification.
+type DeviceNotification struct {
+	DeviceEndpoint DeviceEndpointRef
+	AssetObserver  *AssetCreateObservation
+}
+
+// AssetNotification contains information about an asset notification.
+type AssetNotification struct {
+	Asset         AssetRef
+	DeletionToken AssetDeletionChan
+}
+
+// AssetCreateObservation represents an observation for asset creation events.
+type AssetCreateObservation struct {
+	assetChan chan AssetNotification
+}
+
+// FileMountMap tracks device endpoints and their associated assets.
+type FileMountMap struct {
+	mu             sync.RWMutex
+	fileMountPath  string
+	deviceMap      map[string]DeviceData
+	createDeviceTx chan DeviceNotification
+	logger         *slog.Logger
+}
+
+// DeviceData contains data related to a device endpoint.
+type DeviceData struct {
+	assetCreationTx chan AssetNotification
+	trackedAssets   map[string]AssetDeletionChan
+}
+
+// NewFileMountMap creates a new FileMountMap.
+func NewFileMountMap(fileMountPath string, logger *slog.Logger) *FileMountMap {
+	return &FileMountMap{
+		fileMountPath:  fileMountPath,
+		deviceMap:      make(map[string]DeviceData),
+		createDeviceTx: make(chan DeviceNotification, 100),
+		logger:         logger,
+	}
+}
+
+// UpdateDeviceEndpoints updates the device endpoints in the file mount map.
+func (f *FileMountMap) UpdateDeviceEndpoints(
+	devices map[string]DeviceEndpointRef,
+) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	for deviceKey := range f.deviceMap {
```

Comment:
nit: Consider using the `for deviceKey, data := range f.deviceMap {` syntax. Should save you the `if data, ok` below.